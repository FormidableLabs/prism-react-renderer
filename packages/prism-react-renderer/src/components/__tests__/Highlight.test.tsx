import { render, cleanup } from "@testing-library/react"
import { Highlight } from "../../index"
import { expect } from "vitest"

const exampleCode = `
(function someDemo() {
  var test = "Hello World!";
  console.log(test);
})();

return () => <App />;
`.trim()
describe("<Highlight />", () => {
  afterEach(cleanup)
  describe("snapshots", () => {
    it("renders correctly", () => {
      const { container } = render(
        <Highlight code={exampleCode} language="jsx">
          {({ className, style, tokens, getLineProps, getTokenProps }) => (
            <pre className={className} style={style}>
              {tokens.map((line, i) => (
                <div
                  key={i}
                  {...getLineProps({
                    line,
                  })}
                >
                  {line.map((token, key) => (
                    <span
                      key={key}
                      {...getTokenProps({
                        token,
                      })}
                    />
                  ))}
                </div>
              ))}
            </pre>
          )}
        </Highlight>
      )
      expect(container).toMatchSnapshot()
    })
    it("renders unsupported languages correctly", () => {
      const { container } = render(
        <Highlight
          code={exampleCode}
          // This is an intentional error to test invalid languages
          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
          // @ts-ignore
          language="abcdefghijklmnop"
        >
          {({ className, style, tokens, getLineProps, getTokenProps }) => (
            <pre className={className} style={style}>
              {tokens.map((line, i) => (
                <div
                  key={i}
                  {...getLineProps({
                    line,
                  })}
                >
                  {line.map((token, key) => (
                    <span
                      key={key}
                      {...getTokenProps({
                        token,
                      })}
                    />
                  ))}
                </div>
              ))}
            </pre>
          )}
        </Highlight>
      )
      expect(container).toMatchSnapshot()
    })
    it("renders a default theme when no theme is passed", () => {
      const { container } = render(
        <Highlight code={exampleCode} language="jsx">
          {({ className, style, tokens, getLineProps, getTokenProps }) => (
            <pre className={className} style={style}>
              {tokens.map((line, i) => (
                <div
                  key={i}
                  {...getLineProps({
                    line,
                  })}
                >
                  {line.map((token, key) => (
                    <span
                      key={key}
                      {...getTokenProps({
                        token,
                      })}
                    />
                  ))}
                </div>
              ))}
            </pre>
          )}
        </Highlight>
      )
      expect(container.innerHTML.includes("style")).toBeTruthy()
    })
  })
  describe("getLineProps", () => {
    it("transforms lineProps inputs correctly", () => {
      const input = {
        style: {
          cursor: "pointer",
        },
        className: "line-class",
        line: [
          {
            types: ["punctuation"],
            content: "!",
          },
        ],
        restPropsTest: true,
      }
      render(
        <Highlight code={exampleCode} language="jsx">
          {({ getLineProps }) => {
            const output = getLineProps(input)
            expect(output).toEqual({
              style: {
                cursor: "pointer",
                color: expect.any(String),
              },
              className: "token-line line-class",
              restPropsTest: true,
            })
            return <div></div>
          }}
        </Highlight>
      )
    })
  })
  describe("getTokenProps", () => {
    it("transforms tokenProps inputs correctly", () => {
      const input = {
        style: {
          cursor: "pointer",
        },
        className: "token-class",
        token: {
          types: ["keyword"],
          content: "function",
        },
        restPropsTest: true,
      }
      render(
        <Highlight code={exampleCode} language="jsx">
          {({ getTokenProps }) => {
            const output = getTokenProps(input)
            expect(output).toEqual({
              style: {
                cursor: "pointer",
                color: expect.any(String),
              },
              className: "token keyword token-class",
              restPropsTest: true,
              children: "function",
            })
            return <div></div>
          }}
        </Highlight>
      )
    })
    it("transforms constructor token style correctly", () => {
      // From https://github.com/FormidableLabs/prism-react-renderer/issues/11
      render(
        <Highlight code={"open Common;"} language="reason">
          {({ tokens, getTokenProps }) => {
            const line = tokens[0]
            const token = line[2]
            const output = getTokenProps({
              token,
            })
            expect(typeof output.style).not.toBe("function")
            return <div></div>
          }}
        </Highlight>
      )
    })
  })
})
