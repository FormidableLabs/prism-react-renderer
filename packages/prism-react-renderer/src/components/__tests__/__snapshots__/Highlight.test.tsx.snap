// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Highlight /> > snapshots > renders correctly 1`] = `
<div>
  <pre
    class="prism-code language-jsx"
    style="color: #9CDCFE; background-color: #1E1E1E;"
  >
    <div
      class="token-line"
      style="color: #9CDCFE;"
    >
      <span
        class="token punctuation"
        style="color: rgb(212, 212, 212);"
      >
        (
      </span>
      <span
        class="token keyword"
        style="color: rgb(86, 156, 214);"
      >
        function
      </span>
      <span
        class="token plain"
      >
         
      </span>
      <span
        class="token function"
        style="color: rgb(220, 220, 170);"
      >
        someDemo
      </span>
      <span
        class="token punctuation"
        style="color: rgb(212, 212, 212);"
      >
        (
      </span>
      <span
        class="token punctuation"
        style="color: rgb(212, 212, 212);"
      >
        )
      </span>
      <span
        class="token plain"
      >
         
      </span>
      <span
        class="token punctuation"
        style="color: rgb(212, 212, 212);"
      >
        {
      </span>
      <span
        class="token plain"
      />
    </div>
    <div
      class="token-line"
      style="color: #9CDCFE;"
    >
      <span
        class="token plain"
      >
          
      </span>
      <span
        class="token keyword"
        style="color: rgb(86, 156, 214);"
      >
        var
      </span>
      <span
        class="token plain"
      >
         test 
      </span>
      <span
        class="token operator"
        style="color: rgb(212, 212, 212);"
      >
        =
      </span>
      <span
        class="token plain"
      >
         
      </span>
      <span
        class="token string"
        style="color: rgb(206, 145, 120);"
      >
        "Hello World!"
      </span>
      <span
        class="token punctuation"
        style="color: rgb(212, 212, 212);"
      >
        ;
      </span>
      <span
        class="token plain"
      />
    </div>
    <div
      class="token-line"
      style="color: #9CDCFE;"
    >
      <span
        class="token plain"
      >
          
      </span>
      <span
        class="token console class-name"
        style="color: rgb(78, 201, 176);"
      >
        console
      </span>
      <span
        class="token punctuation"
        style="color: rgb(212, 212, 212);"
      >
        .
      </span>
      <span
        class="token method function property-access"
        style="color: rgb(220, 220, 170);"
      >
        log
      </span>
      <span
        class="token punctuation"
        style="color: rgb(212, 212, 212);"
      >
        (
      </span>
      <span
        class="token plain"
      >
        test
      </span>
      <span
        class="token punctuation"
        style="color: rgb(212, 212, 212);"
      >
        )
      </span>
      <span
        class="token punctuation"
        style="color: rgb(212, 212, 212);"
      >
        ;
      </span>
      <span
        class="token plain"
      />
    </div>
    <div
      class="token-line"
      style="color: #9CDCFE;"
    >
      <span
        class="token plain"
      />
      <span
        class="token punctuation"
        style="color: rgb(212, 212, 212);"
      >
        }
      </span>
      <span
        class="token punctuation"
        style="color: rgb(212, 212, 212);"
      >
        )
      </span>
      <span
        class="token punctuation"
        style="color: rgb(212, 212, 212);"
      >
        (
      </span>
      <span
        class="token punctuation"
        style="color: rgb(212, 212, 212);"
      >
        )
      </span>
      <span
        class="token punctuation"
        style="color: rgb(212, 212, 212);"
      >
        ;
      </span>
      <span
        class="token plain"
      />
    </div>
    <div
      class="token-line"
      style="color: #9CDCFE;"
    >
      <span
        class="token plain"
        style="display: inline-block;"
      >
        

      </span>
    </div>
    <div
      class="token-line"
      style="color: #9CDCFE;"
    >
      <span
        class="token plain"
      />
      <span
        class="token keyword control-flow"
        style="color: rgb(86, 156, 214);"
      >
        return
      </span>
      <span
        class="token plain"
      >
         
      </span>
      <span
        class="token punctuation"
        style="color: rgb(212, 212, 212);"
      >
        (
      </span>
      <span
        class="token punctuation"
        style="color: rgb(212, 212, 212);"
      >
        )
      </span>
      <span
        class="token plain"
      >
         
      </span>
      <span
        class="token arrow operator"
        style="color: rgb(212, 212, 212);"
      >
        =&gt;
      </span>
      <span
        class="token plain"
      >
         
      </span>
      <span
        class="token tag punctuation"
        style="color: rgb(212, 212, 212);"
      >
        &lt;
      </span>
      <span
        class="token tag class-name"
        style="color: rgb(78, 201, 176);"
      >
        App
      </span>
      <span
        class="token tag"
        style="color: rgb(78, 201, 176);"
      >
         
      </span>
      <span
        class="token tag punctuation"
        style="color: rgb(212, 212, 212);"
      >
        /&gt;
      </span>
      <span
        class="token punctuation"
        style="color: rgb(212, 212, 212);"
      >
        ;
      </span>
    </div>
  </pre>
</div>
`;

exports[`<Highlight /> > snapshots > renders unsupported languages correctly 1`] = `
<div>
  <pre
    class="prism-code language-abcdefghijklmnop"
    style="color: #9CDCFE; background-color: #1E1E1E;"
  >
    <div
      class="token-line"
      style="color: #9CDCFE;"
    >
      <span
        class="token plain"
      >
        (function someDemo() {
      </span>
    </div>
    <div
      class="token-line"
      style="color: #9CDCFE;"
    >
      <span
        class="token plain"
      >
          var test = "Hello World!";
      </span>
    </div>
    <div
      class="token-line"
      style="color: #9CDCFE;"
    >
      <span
        class="token plain"
      >
          console.log(test);
      </span>
    </div>
    <div
      class="token-line"
      style="color: #9CDCFE;"
    >
      <span
        class="token plain"
      >
        })();
      </span>
    </div>
    <div
      class="token-line"
      style="color: #9CDCFE;"
    >
      <span
        class="token plain"
        style="display: inline-block;"
      >
        

      </span>
    </div>
    <div
      class="token-line"
      style="color: #9CDCFE;"
    >
      <span
        class="token plain"
      >
        return () =&gt; &lt;App /&gt;;
      </span>
    </div>
  </pre>
</div>
`;
